README.md
# Klappa
# What is Klappa?
The main concept of Klappa is to have one or more action bars with hided buttons which will pop up. If you start Klappa the first time, you will see just one button in the center of the screen. This is your first bar and button togehter. 

# How to set up Klappa?
Every bar of Klappa consists of one or more main buttons which are always visible. You can extend every main button with as many popup buttons as you want. To open the option menu just write "/klappa" (/kl) or use the Klappa menu inside the Blizzard interface menu.

In the main options you can hide the default wow mainbar, add or delete a bar and open a frame with all button ids and how they are assigned.

# What are button ids?
In Wow you have 120 button ids. Nearly every action bar addon uses these ids internally, but you dont see them. If two buttons have the same id, they will to the same action if you click them. With Klappa you are free to assign to your buttons every id you want.


CHANGES.txt
----------------------
- fixes for TBC
- update supported TBC Version

KlappaBar.lua
if not (Klappa2) then return; end;
local _G = getfenv(0);

local Klappa2 = Klappa2;
local config = {};

local L = LibStub("AceLocale-3.0"):GetLocale("Klappa2")
local LBF = LibStub("LibButtonFacade",true)


Klappa2.Bar = {}
Klappa2.Bar.prototype = {}
Klappa2.Bar.metatable = {__index = Klappa2.Bar.prototype}


function Klappa2.Bar:new(index)
	local instance = setmetatable({}, self.metatable)
	instance:init(index)
	return instance
end

function Klappa2.Bar.prototype:init(index)
		config = Klappa2.config.bars;
		self.index = index;
		self.locked = true;
		self:LoadDefaults();
		self:AddOptions();
		self:CreateBar();
		self.root.headers = {};
		self.root.headersCount = 0;
		if LBF then
			self.root.LBFGroup = LBF:Group("Klappa2", L["Bar "]..self.index)
			LBF:RegisterSkinCallback("Klappa2", self.SkinChanged, self);
		end
		if(config[self.index].headers == nil) then
			config[self.index].headers = {};
			config[self.index].headers[1] = {};
			config[self.index].headers[1].popups = {};
			self.root.headers[1] = Klappa2.Header:new(1, self);
		else
			for i, button in pairs (config[self.index].headers) do
				self.root.headersCount = self.root.headersCount + 1
				self.root.headers[i] = Klappa2.Header:new(i, self);
			end
		end
		self:LoadPosition();
		self:UpdateLayout();
		
end

function Klappa2.Bar.prototype:CreateBar()
	local name = "Klappa2Bar"..self.index
	self.root = CreateFrame("Button", name, UIParent, "SecureHandlerStateTemplate");
	self.root:SetPoint("CENTER", UIParent, "CENTER", 0, 0);
	self.root:RegisterForDrag("LeftButton");
	self.root:EnableMouse(false);
	self.root:SetMovable(true);
	self.root:SetClampedToScreen(true);
	self.root.name = name;
	
	self.root:SetWidth(config[self.index].size);
	self.root:SetHeight(config[self.index].size);
	self.root:SetScale(config[self.index].buttonScale)
---
--Um die Bar zu sehen:
	-- self.root:SetBackdrop({
		-- bgFile = "Interface\\Tooltips\\UI-Tooltip-Background",
		-- tile = true,
		-- tileSize = 1,
		-- edgeFile = "Interface\\Tooltips\\UI-Tooltip-Border",
		-- edgeSize = 0,
		-- insets = {left = 0, right = 0, top = 0, bottom = 0}
	-- })
	-- self.root:SetBackdropColor(1, 1, 1, 1)
	-- self.root:SetBackdropBorderColor(0.5, 0.5, 0, 0)
----
	self.root.texture = self.root:CreateTexture();
	self.root.texture:SetTexture(0,0,0.5,0);
	self.root.texture:SetAllPoints(self.root);
	
	
	--Rahmen um die Bar zu verschieben
	self.overlay = CreateFrame("Button", name .. "Overlay", self.root, BackdropTemplateMixin and "BackdropTemplate")
	self.overlay:SetPoint("CENTER", self.root, "CENTER")
	self.overlay:SetFrameLevel(self.root:GetFrameLevel()+20)
	--10Pkte größer als die Leiste
	self.overlay:SetWidth(self.root:GetWidth()+10);
	self.overlay:SetHeight(self.root:GetHeight()+10);
	self.overlay:EnableMouse(true)
	self.overlay:RegisterForDrag("LeftButton")
	self.overlay:RegisterForClicks("LeftButtonUp")
	self.overlay:SetBackdrop({
		bgFile = "Interface\\Tooltips\\UI-Tooltip-Background",
		tile = true,
		tileSize = 1,
		edgeFile = "Interface\\Tooltips\\UI-Tooltip-Border",
		edgeSize = 0,
		insets = {left = 0, right = 0, top = 0, bottom = 0}
	})
	self.overlay:SetBackdropColor(0, 1, 0, 0.5)
	self.overlay:SetBackdropBorderColor(0.5, 0.5, 0, 0)
	self.overlay.Text = self.overlay:CreateFontString(nil, "ARTWORK")
	self.overlay.Text:SetFontObject(GameFontNormal)
	self.overlay.Text:SetText("Bar:"..self.index)
	self.overlay.Text:Show()
	self.overlay.Text:ClearAllPoints()
	self.overlay.Text:SetPoint("CENTER", self.overlay, "CENTER")
	self.overlay:Hide()
	--self.root:Show()
	
end

function Klappa2.Bar.prototype:SkinChanged(SkinID, Gloss, Backdrop, Group, Button, Colors)
	if(Group == L["Bar "]..self.index) then
		config[self.index].skin.ID = SkinID
		config[self.index].skin.Gloss = Gloss
		config[self.index].skin.Backdrop = Backdrop
		config[self.index].skin.Colors = Colors
	end
end

function Klappa2.Bar.prototype:UpdateLayout()
	local rootx, rooty = 0, 0;
	local isVert, isRtDn = false, false;
	local size = config[self.index].size;
	local orientation = config[self.index].orient;
	if (orientation == "horzdown") then
		isRtDn = true;
	elseif (orientation == "vertleft") then
		isVert = true;
	elseif (orientation == "vertright") then
		isVert = true;
		isRtDn = true;
	end
	if (isVert) then
		rootx = size;
	else
		rooty = size;
	end
	
	for idx, buttonClass in pairs(self.root.headers) do
		local x, y = 0, 0;
		local padding = config[self.index].padding;
		--Irgendwoher kam ein Fehler, dass "padding" ein table ist
		--hier wird das geprüft und korrigiert
		if (type(padding) == "table") then
			padding = 1 
			config[self.index].padding = padding
		end
		if (isVert) then
			y = (-size-padding)*(buttonClass.index - 1);
			rooty = rooty + size + padding;
			self.root:SetWidth(size)
			self.root:SetHeight(size*self.root.headersCount + (padding * (self.root.headersCount-1)))
			
		else
			x = (size+padding)*(buttonClass.index - 1);
			rootx = rootx + size + padding;
			
			self.root:SetWidth(size*self.root.headersCount + (padding * (self.root.headersCount-1)))
			self.root:SetHeight(size)
			
		end
		buttonClass:UpdateLayout(x, y, isVert, isRtDn);
	end
	
	self.overlay:SetWidth(self.root:GetWidth()+10);
	self.overlay:SetHeight(self.root:GetHeight()+10);
end

function Klappa2.Bar.prototype:ToggleLock()
	if (config[self.index].locked) then
		self.overlay:SetScript("OnDragStart", function() self:StartDrag(); end);
		self.overlay:SetScript("OnDragStop", function() self:StopDrag(); end);
		self.overlay:Show();
		config[self.index].locked = false;
		-- Show all popups
		for idx, buttonClass in pairs(self.root.headers) do
			buttonClass:ShowPops();
		end
	else
		self.overlay:SetScript("OnDragStart", nil);
		self.overlay:SetScript("OnDragStop", nil);
		self.overlay:Hide()
		
		config[self.index].locked = true;
		for idx, buttonClass in pairs(self.root.headers) do
			buttonClass:HidePops();
		end
	end
end

function Klappa2.Bar.prototype:ToggleLockButtons()
	if(config[self.index].lockButtons) then
		config[self.index].lockButtons = false;
	else
		config[self.index].lockButtons = true;
	end
end

function Klappa2.Bar.prototype:StartDrag()
	self.root:StartMoving();
end

function Klappa2.Bar.prototype:StopDrag()
	self.root:StopMovingOrSizing();
	self:SavePosition();
end

function Klappa2.Bar.prototype:LoadPosition()
	local x, y, s = config[self.index].rootX, config[self.index].rootY, self.root:GetEffectiveScale();

	self.root:ClearAllPoints();
	self.root:SetPoint(x and "TOPLEFT" or "CENTER", UIParent, x and "BOTTOMLEFT" or "CENTER", x or 0, y or 0);
end

function Klappa2.Bar.prototype:SavePosition()
	local x, y, s = self.root:GetLeft(), self.root:GetTop(), self.root:GetEffectiveScale();

	config[self.index].rootX = x;
	config[self.index].rootY = y;
end


function Klappa2.Bar.prototype:AddMainButton()
	idx = config[self.index].numberButtons + 1;
	
	config[self.index].headers[idx] = {};
	config[self.index].headers[idx].popups = {};

	self.root.headers[idx] = Klappa2.Header:new(idx, self);
	self.root.headers[idx]:AddPopup()
	config[self.index].numberButtons = idx;
	self.root.headersCount = self.root.headersCount + 1
	self:UpdateLayout();
end

function Klappa2.Bar.prototype:DelMainButton()
	index = config[self.index].numberButtons;
	if (index == 0 or index == nil) then return end;
	self.root.headers[index].header:Hide();
	self.root.headers[index].header.popups = nil;
	self.root.headers[index].header = nil;
	self.root.headers[index] = nil;

	config[self.index].headers[index] = nil;
	Klappa2.options.args["Bar"..self.index].args["Header"..index] = nil;
	config[self.index].numberButtons = index-1;
	self.root.headersCount = self.root.headersCount - 1
	self:UpdateLayout();
end

function Klappa2.Bar.prototype:LoadDefaults()
	--if(config[self.index] == nil) then config[self.index] = {} end  --Änderung
	if (config[self.index].numberButtons == nil) then config[self.index].numberButtons = 1; end
	if (config[self.index].orient == nil) then config[self.index].orient = "vertleft"; end
	if (config[self.index].buttonScale == nil) then config[self.index].buttonScale = 1; end
	if (config[self.index].popupScale == nil) then config[self.index].popupScale = 1; end
	if (config[self.index].alpha == nil) then config[self.index].alpha = 1; end
	if (config[self.index].popupAlpha == nil) then config[self.index].popupAlpha = 1; end
	if (config[self.index].locked == nil) then config[self.index].locked = true; end
	if (config[self.index].lockButtons == nil) then config[self.index].lockButtons = true; end
	if (config[self.index].padding == nil) then config[self.index].padding = 1; end
	if (config[self.index].size == nil) then config[self.index].size = 36; end
	if (config[self.index].tooltip == nil) then config[self.index].tooltip = true; end
	if (config[self.index].skin == nil) then config[self.index].skin = {}; end
end

function Klappa2.Bar.prototype:AddOptions()
	Klappa2.options.args["Bar"..self.index] = {
		name = L["Bar "]..self.index,
		desc = L["Options for the bar"],
		type = "group",
		args = {}
		};
	Klappa2.options.args["Bar"..self.index].args = {
		lock = {
			name = L["Lock the bar"],
			desc = L["Lock the bar"],
			type = "toggle",
			order = 1,
			get = function() return config[self.index].locked end,
			set = function(info,value) self:ToggleLock() end,
			disabled = function() return InCombatLockdown() end,
		},
		lockbuttons = {
			name = L["Button Lock"],
			desc = L["Lock the buttons"],
			type = "toggle",
			order = 2,
			get = function() return config[self.index].lockButtons end,
			set = function(info,value) self:ToggleLockButtons() end,
		},
		tooltip = {
			name = L["Tooltip"],
			desc = L["Enable/disable the tooltips"],
			type = "toggle",
			order = 3,
			get = function() return config[self.index].tooltip end,
			set = function(info,value) config[self.index].tooltip = value end,
		},
		padding = {
			type = "range",
			name = L["Padding"],
			desc = L["Set the padding of the buttons"],
			order = 4,
			step = 0.05,
			min = 0.0,
			max = 25.0,
			isPercent = false,
			get = function() return config[self.index].padding end,
			set = function(info,value)
				config[self.index].padding = value;
				self:UpdateLayout();
			end
		},
		mainScale = {
			type = "range",
			name = L["Scale buttons"],
			desc = L["Scale the buttons"],
			order = 5,
			step = 0.05,
			min = 0.25,
			max = 5.0,
			isPercent = false,
			get = function() return config[self.index].buttonScale; end,
			set = function(info,value)
				config[self.index].buttonScale = value; self.root:SetScale(value); end
		},
		orientation = {
			type = "select",
			name = L["Orientation"],
			desc = L["Set the orientation of the Klappa2 Bar."],
			order = 6,
			get = function() return config[self.index].orient; end,
			set = function(info,value) config[self.index].orient = value; self:UpdateLayout(); end,
			values = {
				["horzup"] = L["Horizontal, Grow Up"],
				["horzdown"] = L["Horizontal, Grow Down"],
				["vertright"] = L["Vertical, Grow Right"],
				["vertleft"] = L["Vertical, Grow Left"],
			},
		},
		add = {
			name = L["New main button"],
			desc = L["Add a new main button"],
			order = 7,
			type = "execute",
			func = function() self:AddMainButton() end,
		},
		del = {
			name = L["Delete main button"],
			desc = L["Delete the last main button"],
			order = 8,
			type = "execute",
			func = function() self:DelMainButton() end,
		},
	}
end


locale/deDE.lua
local L = LibStub("AceLocale-3.0"):NewLocale("Klappa2","deDE", false)
if not L then return end

	L["Hide mainbar"] = "Standard Buttonleiste verstecken"
	L["Hides the default mainbar"] = "Versteckt die Standard Buttonleiste"
	L["Button Lock"] = "Kn\195\182pfe sperren"
	L["Lock the buttons"] = "Die Kn\195\182pfe sperren"
	L["Lock the bar"] = "Leiste sperren"
	L["Scale buttons"] = "Kn\195\182pfe skalieren"
	L["Scale the buttons"] = "Buttons skalieren"
	L["Invert popup direction"] = "Popuprichtung invertieren"
	L["Vertical bar direction"] = "Vertikale Balkenrichtung"
	L["Tooltip configuration"] = "Tooltip Einstellungen"
	L["Show Tooltip"] = "Tooltip anzeigen"
	L["Out of combat only"] = "Nur au\195\159erhalb des Kampfes"
	L["Klappa2 configuration"] = "Klappa2bar Einstellungen"
	L["Popup Button Scale"] = "Popup Kn\195\182pfe Skalierung"
	L["Scale the popup buttons"] = "Popup Kn\195\182pfe skalieren"
	L["Button Style"] = "Aussehen der Kn\195\182pfe"
	L["Show short tooltip"] = "Kurzer Tooltiptext"
	L["Orientation"] = "Ausrichtung"
	L["Set the orientation of the Klappa2 Bar."] = "Ver\195\164ndert die Ausrichtung der Klappa2 Bar"
	L["Horizontal, Grow Up"] = "Horizontal, nach oben"
	L["Horizontal, Grow Down"] = "Horizontal, nach unten" 
	L["Vertical, Grow Right"] = "Vertikal, nach rechts" 
	L["Vertical, Grow Left"] = "Vertikal, nach links" 
	L["Add popup button"] = "Popupbutton hinzuf\195\188gen" 
	L["Add a new popup button"] = "Einen neuen Popupbutton hinzuf\195\188gen" 
	L["New row"] = "Neue Reihe" 
	L["Add a new row"] = "Eine neue Reihe hinzuf\195\188gen" 
	L["New main button"] = "Neuer Hauptbutton" 
	L["Add a new main button"] = "Einen neuen Hauptbutton hinzuf\195\188gen" 
	L["Delete main button"] = "L\195\182sche Hauptbutton"	
	L["Delete the last main button"] = "L\195\182scht den letzten Hauptbutton"
	L["Main button "] = "Hauptbutton "
	L["Options for this main button"] = "Einstellungen f\195\188r diesen Hauptbutton"
	L["Row "] = "Reihe " 
	L["Options for this row"] = "Einstellungen f\195\188r diese Reihe" 
	L["Delete row"] = "L\195\182sche Reihe" 
	L["Delete last row"] = "L\195\182scht die letzte Reihe" 
	L["Delete popup button"] = "L\195\182sche Popupbutton" 
	L["Delete the last popup button"] = "L\195\182sche den letzten Popupbutton" 
	L["Tooltip"] = "Tooltip" 
	L["Enable/disable the tooltips"] = "Aktivieren/deaktivieren der Tooltips" 
	L["Padding"] = "Zwischenabstand" 
	L["Set the padding of the buttons"] = "Den Zwischenabstand der Buttons anpassen" 
	L["Change main button-ID"] = "Hauptbutton-ID \195\164ndern" 
	L["Change the ID of this mainbutton"] = "Die ID des Hauptbuttons \195\164ndern" 
	L["Bar "] = "Leiste " 
	L["Options for the bar"] = "Einstellungen f\195\188r diese Leiste" 
	L["Add Bar"] = "Leiste hinzuf\195\188gen" 
	L["Add a new bar"] = "Eine neue Leiste hinzuf\195\188gen" 
	L["Delete Bar"] = "Leiste l\195\182schen" 
	L["Delete the last bar"] = "Die letzte Leiste l\195\182schen" 
	L["Popup button "] = "Popupbutton " 
	L["Options for the popup button"] = "Einstellungen f\195\188r den Popupbutton" 
	L["Change the ID of this popup button"] = "Die ID von diesem Popupbutton \195\164ndern" 
	L["Change ID of popup button: "] = "ID \195\164ndern von Popupbutton: " 
	L["Show all buttonids"] = "Zeige alle Button-IDs" 
	L["Shows all buttons with their ids"] = "Zeige alle Buttons mit den entsprechenden IDs" 
	L["no more popup button to delete"] = "Keine Popup-Button mehr zum L\195\182schen"
	L["Left Click|r to open configuration"] = "Links Klick|r für das Konfigurationsmen\195\188"
	L["Right Click|r to lock/unlock bars"] = "Rechts Klick|r zum Sperren/Entsperren der Leisten"


locale/enEN.lua
local L = LibStub("AceLocale-3.0"):NewLocale("Klappa2","enUS",true)
if not L then return end

	L["Hide mainbar"] = true 	
	L["Hides the default mainbar"] = true 
	L["Button Lock"] = true 
	L["Lock the buttons"] = true 
	L["Lock the bar"] = true 
	L["Scale buttons"] = true 
	L["Scale the buttons"] = true 
	L["Invert popup direction"] = true 
	L["Vertical bar direction"] = true 
	L["Tooltip configuration"] = true 
	L["Show Tooltip"] = true 
	L["Out of combat only"] = true 
	L["Klappa2 configuration"] = true 
	L["Popup Button Scale"] = true 
	L["Scale the popup buttons"] = true 
	L["Button Style"] = true 
	L["Show short tooltip"] = true 
	L["Orientation"] = true 
	L["Set the orientation of the Klappa2 Bar."] = true 
	L["Horizontal, Grow Up"] = true 
	L["Horizontal, Grow Down"] = true
	L["Vertical, Grow Right"] = true 
	L["Vertical, Grow Left"] = true 
	L["Add popup button"] = true 
	L["Add a new popup button"] = true 
	L["New row"] = true 
	L["Add a new row"] = true 
	L["New main button"] = true 
	L["Delete main button"] = true	
	L["Delete the last main button"] = true
	L["Add a new main button"] = true 
	L["Main button "] = true
	L["Options for this main button"] = true
	L["Row "] = true 
	L["Options for this row"] = true 
	L["Delete row"] = true 
	L["Delete last row"] = true 
	L["Delete popup button"] = true 
	L["Delete the last popup button"] = true 
	L["Tooltip"] = true 
	L["Enable/disable the tooltips"] = true 
	L["Padding"] = true 
	L["Set the padding of the buttons"] = true 
	L["Change main button-ID"] = true 
	L["Change the ID of this mainbutton"] = true 
	L["Bar "] = true 
	L["Options for the bar"] = true 
	L["Add Bar"] = true 
	L["Add a new bar"] = true 
	L["Delete Bar"] = true 
	L["Delete the last bar"] = true 
	L["Popup button "] = true 
	L["Options for the popup button"] = true 
	L["Change the ID of this popup button"] = true 
	L["Change ID of popup button: "] = true 
	L["Show all buttonids"] = true 
	L["Shows all buttons with their ids"] = true
	L["no more popup button to delete"] = true
	L["Left Click|r to open configuration"] = true
	L["Right Click|r to lock/unlock bars"] = true


Klappa.lua
Klappa2 = LibStub("AceAddon-3.0"):NewAddon("Klappa2", "AceConsole-3.0")  --"LibFuBarPlugin-3.0",

local _G = getfenv(0)
local L = LibStub("AceLocale-3.0"):GetLocale("Klappa2")

-- LDB
local ldb = LibStub:GetLibrary("LibDataBroker-1.1"):NewDataObject("Klappa", {
    type = "launcher",
    icon = "Interface\\Icons\\INV_Weapon_ShortBlade_17",
    OnClick = function(self, button)
		if (button == "RightButton") then
			for idx, bar in pairs(Klappa2.bars) do
				bar:ToggleLock()
			end
		else
			LibStub("AceConfigDialog-3.0"):Open("Klappa")
		end
	end,
	OnTooltipShow = function(Tip)
		if not Tip or not Tip.AddLine then
			return
		end
		Tip:AddLine("Klappa")
		Tip:AddLine("|cFFff4040"..L["Left Click|r to open configuration"], 1, 1, 1)
		Tip:AddLine("|cFFff4040"..L["Right Click|r to lock/unlock bars"], 1, 1, 1)
	end,
})
--------------------------------------

-- Fubar-Options
Klappa2.name = "Klappa";
Klappa2.hasIcon = "Interface\\Icons\\INV_Weapon_ShortBlade_17";
Klappa2.hasNoColor = true;
Klappa2.defaultMinimapPosition = 200;
Klappa2.clickableTooltip  = false;
Klappa2.independentProfile = true;
Klappa2.defaultPosition = "RIGHT";
Klappa2.hideWithoutStandby = true;
Klappa2.cannotDetachTooltip = true;

--Fubar methods
function Klappa2:OnUpdateFuBarText()
	self:SetFuBarText("Klappa")
end

function Klappa2:OnUpdateFuBarTooltip()
    GameTooltip:AddLine("Klappa")
    GameTooltip:AddLine("Click to open the configuration", 0, 1, 0)
end

function Klappa2:OnFuBarClick(button)
	self:OpenConfigMenu();
end
--------------------------------

local defaults =
{
	char =
	{
	bars = {
		{
			headers = {
				{
					popups =
					{
						{
							["id"] = 1,
						}, -- [1]
					},
				},
			},
			skin = {},
			orient = "vertleft",
			buttonScale = 1,
			popupScale = 1,
			alpha = 1,
			popupAlpha = 1,
			locked = true,
			lockButtons = true,
			padding = 1,
			size = 35,
			tooltip = true,
			numberButtons = 1,
		},
	},

	hideUI = false,
	numberBars = 1
	}
}

----------------
function Klappa2:OnInitialize()

	self.db = LibStub("AceDB-3.0"):New("KlappaDB", defaults)
end

function Klappa2:InitOptions()
	local options = {
		name = "Klappa",
		desc = "Action bars with popup buttons",
		icon = "Interface\\Icons\\INV_Weapon_ShortBlade_17",
		type="group",
		args = {
			showUI = {
				name = L["Hide mainbar"],
				desc = L["Hides the default mainbar"],
				type = "toggle",
				order = 1,
				get = function() return Klappa2.config.hideUI end,
				set = function(info,value) Klappa2:SetDefaultUIElements(value); Klappa2.config.hideUI = value; end,
			},

			add = {
				name = L["Add Bar"],
				desc = L["Add a new bar"],
				type = "execute",
				order = 3,
				func = function() self:AddBar() end,
			},
			del = {
				name = L["Delete Bar"],
				desc = L["Delete the last bar"],
				type = "execute",
				order = 6,
				func = function() self:DeleteBar() end,
			},
			ids = {
				name = L["Show all buttonids"],
				desc = L["Shows all buttons with their ids"],
				type = "execute",
				order = 9,
				func = function() self:ShowIDs() end,
			},
		}
	}

	return options;
end

function Klappa2:OnEnable()
	Klappa2.config = self.db.char;
	self.options = self:InitOptions();
	self.OnMenuRequest = self.options;
	--self:SetConfigTable(self.options);	--for Fubar  config
	--self:SetFuBarIcon("Interface\\Icons\\INV_Weapon_ShortBlade_17");
	--self:ToggleFuBarMinimapAttached()
	--self:Hide()
	self:CreateRoot();
	self:CreateShowIDs()
	self:SetDefaultUIElements(Klappa2.config.hideUI)
	--altes self:SetConfigTable(self.options);

	LibStub("AceConfig-3.0"):RegisterOptionsTable("Klappa", self.options)
	self.optionsFrame = LibStub("AceConfigDialog-3.0"):AddToBlizOptions("Klappa", "Klappa")
	self.optionsFrameGui = LibStub("AceConfigDialog-3.0"):Open("Klappa")
	self:RegisterChatCommand("kl", "ChatCommand")
	self:RegisterChatCommand("klappa", "ChatCommand")

end

function Klappa2:ChatCommand(input)
	if not input or input:trim() == "" then
	--InterfaceOptionsFrame_OpenToCategory(self.optionsFrame)
	LibStub("AceConfigDialog-3.0"):Open("Klappa")
  else
  print("console")
    LibStub("AceConfigCmd-3.0").HandleCommand(Klappa, "kl2", "Options", input)
  end
end

function Klappa2:SetDefaultUIElements(v)
	if(v) then
		-- Hide the main grafics
		MainMenuBarArtFrame:Hide()
		MainMenuBar:Hide()
	else
		--Show the main grafics
		MainMenuBarArtFrame:Show()
		MainMenuBar:Show()
	end
end

function Klappa2:CreateRoot()
	self.bars = {};

	if (Klappa2.config.bars == nil) then
		self.bars[1] = Klappa2.Bar:new(1);
	else
		for i, bar in pairs (Klappa2.config.bars) do
			self.bars[i] = Klappa2.Bar:new(i);
		end
	end
	self:HideEmpty();
end

function Klappa2:AddBar()
	idx = Klappa2.config.numberBars + 1;
	Klappa2.config.bars[idx] = {};
	self.bars[idx] = Klappa2.Bar:new(idx);
	Klappa2.config.numberBars = idx;
end

function Klappa2:DeleteBar()
	local idx = Klappa2.config.numberBars;
	if (idx == 0 or idx == nil) then return end;
	self.bars[idx].root:Hide();
	for i, mainbutton in pairs (self.bars[idx].root.headers) do
		self.bars[idx]:DelMainButton();
	end

	Klappa2.config.bars[idx] = nil;
	Klappa2.options.args["Bar"..idx] = nil;
	Klappa2.config.numberBars = idx-1;
end

function Klappa2:HideEmpty()
	for k, bar in pairs(Klappa2.bars) do
		for i, main in pairs(bar.root.headers) do
			for k,popup in pairs(bar.root.headers[i].popupButtons) do
				if not ( k == 1 ) then
					local popuptexture = GetActionTexture(popup.button.id);
					if not(popuptexture) then
						popup.button:Hide();
					end
				end
			end
		end
	end
end

--Creating/handling the button id overview

function Klappa2:ShowIDs()
	if(Klappa2.showIDs.show) then
		Klappa2.showIDs.show = false;
		Klappa2.showIDs:Hide();
		Klappa2.showIDs:UnregisterEvent("ACTIONBAR_SLOT_CHANGED");
	else
		Klappa2:UpdateIDs();
		Klappa2.showIDs:Show();
		Klappa2.showIDs.show = true;
		Klappa2.showIDs:RegisterEvent("ACTIONBAR_SLOT_CHANGED");
	end
end

function Klappa2:CreateShowIDs()
	local buttonsize = 35
	Klappa2.showIDs = CreateFrame("Frame","ShowIDs", UIParent);
	Klappa2.showIDs:SetMovable(true);
	Klappa2.showIDs:SetClampedToScreen(true);
	Klappa2.showIDs:SetPoint("CENTER",0,0);
	Klappa2.showIDs:SetHeight((10*buttonsize)+20);
	Klappa2.showIDs:SetWidth(12*buttonsize);

	--Rahmen um das Fenster zu verschieben
	Klappa2.handle = CreateFrame("Button", "ShowIdHandle", Klappa2.showIDs, BackdropTemplateMixin and "BackdropTemplate")
	Klappa2.handle:SetPoint("TOPLEFT", Klappa2.showIDs, "TOPLEFT")
	Klappa2.handle:SetFrameLevel(Klappa2.showIDs:GetFrameLevel()+20)
	Klappa2.handle:SetWidth(12*buttonsize);
	Klappa2.handle:SetHeight(20);
	Klappa2.handle:EnableMouse(true)
	Klappa2.handle:RegisterForDrag("LeftButton")
	Klappa2.handle:SetBackdrop({
		bgFile = "Interface\\Tooltips\\UI-Tooltip-Background",
		tile = true,
		tileSize = 1,
		edgeFile = "Interface\\Tooltips\\UI-Tooltip-Border",
		edgeSize = 0,
		insets = {left = 0, right = 0, top = 0, bottom = 0}
	})
	Klappa2.handle:SetBackdropColor(0, 1, 1, 0.5)
	Klappa2.handle:SetBackdropBorderColor(0.5, 0.5, 0, 0)
	Klappa2.handle.text = Klappa2.handle:CreateFontString();
	Klappa2.handle.text:SetFont("Fonts\\FRIZQT__.TTF",15);
	Klappa2.handle.text:SetText("Button IDs");
	Klappa2.handle.text:SetAllPoints(Klappa2.handle);

	Klappa2.handle:SetScript("OnDragStart", function() self:StartDrag(); end);
	Klappa2.handle:SetScript("OnDragStop", function() self:StopDrag(); end);

	Klappa2.showIDs:Hide()
	Klappa2.showIDs.show = false;

	local i = 0;
	Klappa2.showIDsTex = {};
	while (i<120) do
		Klappa2.showIDsTex[i] = CreateFrame("Button","ShowIDsTexture"..i, Klappa2.showIDs);
		row,_ = -math.modf(i/12)
		x = i%12;
		Klappa2.showIDsTex[i]:SetPoint("TOPLEFT",x*buttonsize,(row*buttonsize)-20);
		Klappa2.showIDsTex[i]:SetHeight(buttonsize);
		Klappa2.showIDsTex[i]:SetWidth(buttonsize);
		Klappa2.showIDsTex[i].texture = Klappa2.showIDsTex[i]:CreateTexture();
		Klappa2.showIDsTex[i].texture:SetAllPoints(Klappa2.showIDsTex[i]);
		Klappa2.showIDsTex[i].text = Klappa2.showIDsTex[i]:CreateFontString();
		Klappa2.showIDsTex[i].text:SetFont("Fonts\\FRIZQT__.TTF",15);
		Klappa2.showIDsTex[i].text:SetText(i+1);
		Klappa2.showIDsTex[i].text:SetAllPoints(Klappa2.showIDsTex[i]);
		Klappa2.showIDsTex[i].id = i + 1;
		Klappa2.showIDsTex[i]:HookScript("OnEnter", function(self, motion) Klappa2:ShowTooltip(self, self.id); end);
		Klappa2.showIDsTex[i]:HookScript("OnLeave", function(self, motion) Klappa2:HideTooltip(self); end);

		i=i+1;
	end
	Klappa2.showIDs:SetScript("OnEvent", function(self, event, id) Klappa2:Update(id); end);
end

function Klappa2:Update(id)
	if (Klappa2.showIDs == nil) then return end;
	if(Klappa2.showIDs.show) then
		if(id == 0) then UpdateIDs(); return; end
		local texture = GetActionTexture(id);
		Klappa2.showIDsTex[id-1].texture:SetTexture(texture);
	end
end

function Klappa2:ShowTooltip(button, id)
	GameTooltip:SetOwner(button,ANCHOR_RIGHT);
	GameTooltip:SetAction(id);
end

function Klappa2:HideTooltip(button)
	if GameTooltip:IsOwned(button) then
	--print("hide")
		GameTooltip:Hide();
	end
end

function Klappa2:UpdateIDs()
	local i = 0;
	while (i<120) do
		local texture = GetActionTexture(i+1);
		Klappa2.showIDsTex[i].texture:SetTexture(texture);
		i = i+1;
	end
end

function Klappa2:StartDrag()
	Klappa2.showIDs:StartMoving();
end

function Klappa2:StopDrag()
	Klappa2.showIDs:StopMovingOrSizing();
end
--------------------------------------------------
